### PR TITLE
Admin reporting view

### LINKED ISSUE #1926: Metrics Dashboard
Better understand the usage and impact of CiviForm.

Full doc [here](https://docs.google.com/document/d/1vg_s5LsRqcZ-ZhH0zfV5PnxhINRuUhxYlxaEZRa-vDw/edit?resourcekey=0-jWa5yMa2Xe9QHNQna_3NDg#heading=h.f0vy7bg4xle)

### PR DESCRIPTION
**Pull request by bion**
### Description

Implements a reporting page for the admins backed by a postgres materialized view. Feature can be enabled by turning on `ADMIN_REPORTING_UI_ENABLED`.

This is feature relies on the durable job system. It should not be used until the `DurableJobModule` has been enabled.

Screenshot from 2023-02-21 18-08-20

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### New Features

- [x] Add new FeatureFlag env vars to `server/conf/application.conf`
- [x] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)

### Issue(s) this completes

Fixes #1926

### CODE REVIEW

**Review by rockycodes — COMMENTED** [review]
This was helpful to look through, thanks for including me! I don't feel like I have enough context to approve, but the basic setup makes sense to me.

If it doesn't already exist, it would be helpful to do a docs update about how to run Durable Jobs adhoc on dev.

**Review by avaleske — APPROVED** [review]
LGTM!

A few questions through:
- There were a number of places where the presubmits noticed there weren't tests, such as some controllers. Is that intentional?
- I didn't see any a11y affordances on the tooltips

File: server/app/controllers/admin/AdminReportingController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
+  public Result index() {
+    return ok(adminReportingIndexView.render(reportingService.getMonthlyStats()));
+  }
+
+  /** Identifiers for the different data sets available for download. */
+  public enum DataSetName {
+    APPLICATION_COUNTS_BY_PROGRAM,
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/admin/AdminReportingController.java:34]: I haven't found or heard of the need for that with Java enums outside of proto land

File: server/app/services/DateConverter.java [hunk 2]
```diff
@@ -90,4 +93,12 @@ public String formatIso8601Date(LocalDate date) {
   public LocalDate renderLocalDate(Long timestamp) {
     return Instant.ofEpochMilli(timestamp).atZone(this.zoneId).toLocalDate();
   }
+
+  /** Formats a {@link java.sql.Timestamp} to MM/YY. */
+  public String renderMonthAndYear(Timestamp timestamp) {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/services/DateConverter.java:98]: done

File: server/app/views/admin/AdminLayout.java [hunk 3]
```diff
@@ -118,24 +116,51 @@ private NavTag renderNavBar() {
...
+    // Once feature flag is removed for reporting the program admin nav will include
+    // links for programs and reporting.
+    if (primaryAdminType.equals(AdminType.PROGRAM_ADMIN)
+        && !getFeatureFlags().isAdminReportingUiEnabled()) {
+      return adminHeader.with(headerLink("Logout", logoutLink, "float-right"));
+    }
+
+    if (primaryAdminType.equals(AdminType.PROGRAM_ADMIN)) {
+      adminHeader.with(programsHeaderLink).with(reportingHeaderLink);
+    } else {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/AdminLayout.java:154]: There will be quite a few such places to update if we ever do add another admin role. The worst case scenario here is showing links to an admin that shouldn't see them and can't access them.

If you strongly prefer an exhaustive check here lmk and I'll add it.

File: server/app/repository/ReportingRepository.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /** Triggers a refresh of the monthly reporting view. */
+  public void refreshMonthlyReportingView() {
+    database.sqlUpdate("REFRESH MATERIALIZED VIEW monthly_submissions_reporting_view").execute();
+  }
+
+  private static double getSecondsFromPgIntervalRowValue(SqlRow row, String key) {
+    Object interval = row.get(key);
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/repository/ReportingRepository.java:90]: Updated to do null checking inline instead

File: server/app/views/admin/reporting/AdminReportingIndexView.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+      headerRow.with(thTag);
+    }
+
+    TableTag table =
+        table().withClasses("border-collapse table-auto w-full").with(thead(headerRow), tbodyTag);
+
+    int rowCount = 0;
+    for (TrTag dataRow : dataRows) {
+      tbodyTag.with(dataRow.withCondClass(rowCount++ % 2 == 0, "bg-slate-100"));
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/reporting/AdminReportingIndexView.java:221]: Nice! done

File: server/test/repository/ReportingRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                    applicationSpec.getLeft(),
+                    applicationSpec.getMiddle(),
+                    applicationSpec.getRight()));
+
+    repo.refreshMonthlyReportingView();
+
+    assertThat(repo.loadMonthlyReportingView())
+        .containsExactly(
+            ApplicationSubmissionsStat.create(
+                "Fake Program A", getMonthTimestamp(lastMonth), 3L, 300, 500, 750, 990),
```
[the remark this answers is not in the document]
Comment (bion) [reply server/test/repository/ReportingRepositoryTest.java:75]: done

File: server/test/repository/ReportingRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            applicationSpec ->
+                createFakeApplication(
+                    programB,
+                    applicationSpec.getLeft(),
+                    applicationSpec.getMiddle(),
+                    applicationSpec.getRight()));
+
+    assertThat(repo.loadThisMonthReportingData())
+        .containsExactly(
+            ApplicationSubmissionsStat.create(
```
[the remark this answers is not in the document]
Comment (bion) [reply server/test/repository/ReportingRepositoryTest.java:114]: done

**Comment by bion** [comment]
@avaleske 

> * There were a number of places where the presubmits noticed there weren't tests, such as some controllers. Is that intentional?

Yes. The controller and view are exercised by browser tests, which don't show up in codecov.

> * I didn't see any a11y affordances on the tooltips

They're admin-only atm, which we do not have a11y standards for.

File: server/test/repository/ReportingRepositoryTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (avaleske) [reply server/test/repository/ReportingRepositoryTest.java:77]: This is helpful! Do you think a note explaining the math would help as well, something like
```
// The expected values here have submission duration percentile stats calculated from
// the submitted (i.e. active and obsolete) applications. The application submission times
// range from 0 to 1000, so we expect the 50th percentile to be 500.
```
or is that over-explaining?

File: server/app/views/admin/AdminLayout.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (avaleske) [reply server/app/views/admin/AdminLayout.java:154]: makes sense, so I'm good with it as is

**Comment by avaleske** [comment]
Thanks for the changes! I had one question in response to one of them but it's a very soft suggestion since the comment you added helped a lot.

**Pull request merged**